capnp/src/message.rs

```diff
@@ -20,7 +20,54 @@
 // THE SOFTWARE.
 
 //! Untyped root container for a Cap'n Proto value.
-
+//!
+//! ## Notes about type specialization
+//! This module provides [TypedReader] and [TypedBuilder] structs which are strongly-typed variants
+//! of [Reader] and [Builder].
+//!
+//! Code autogenerated by capnpc will have an individual module for each of structures and each of
+//! modules will have `Owned` struct which implements [capnp::traits::Owned] trait.
+//!
+//! Example from a real auto-generated file:
+//!
+//! ```ignore
+//! pub mod simple_struct {
+//!     #[derive(Copy, Clone)]
+//!     pub struct Owned(());
+//!     impl <'a> ::capnp::traits::Owned<'a> for Owned { type Reader = Reader<'a>; type Builder = Builder<'a>; }
+//!     ....
+//! }
+//! ```
+//!
+//! [TypedReader] and [TypedBuilder] accept generic type parameter `T`. This parameter must be
+//! a corresponding `Owned` type which was auto-generated inside the corresponding module.
+//!
+//! For example, for auto-generated module `crate::test_data::simple_struct` you'd supply 
+//! `crate::test_data::simple_struct::Owned` type into [TypedReader]/[TypedBuilder]
+//!
+//! ```ignore
+//! include!(concat!(env!("OUT_DIR"), "/simple_struct_capnp.rs"));
```

Comment:
Directly including the generated code like this is not guaranteed to work. The usual pattern is to include the code inside a `mod`: https://github.com/capnproto/capnproto-rust/commit/42a460e6615f0889f237912daa5e4b1627a65e7e

Just now I put up a proposal for a macro that makes this less verbose: https://github.com/capnproto/capnproto-rust/pull/573